cloud/services/compute/loadbalancers/reconcile_test.go

```diff
@@ -0,0 +1,167 @@
+	_ = clusterv1.AddToScheme(scheme.Scheme)
+	_ = infrav1.AddToScheme(scheme.Scheme)
+}
+
+var fakeBootstrapSecret = &corev1.Secret{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "my-cluster-bootstrap",
+		Namespace: "default",
+	},
+	Data: map[string][]byte{
+		"value": []byte("Zm9vCg=="),
+	},
+}
+
+var fakeCluster = &clusterv1.Cluster{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "my-cluster",
+		Namespace: "default",
+	},
+	Spec: clusterv1.ClusterSpec{},
+}
+
+var fakeGCPCluster = &infrav1.GCPCluster{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "my-cluster",
+		Namespace: "default",
+	},
+	Spec: infrav1.GCPClusterSpec{
+		Project: "my-proj",
+		Region:  "us-central1",
+	},
+	Status: infrav1.GCPClusterStatus{
+		FailureDomains: clusterv1.FailureDomains{
+			"us-central1-a": clusterv1.FailureDomainSpec{ControlPlane: true},
+		},
+	},
+}
+
+func TestService_createOrGetInstanceGroup(t *testing.T) {
+	fakec := fake.NewClientBuilder().
+		WithScheme(scheme.Scheme).
+		Build()
+
+	clusterScope, err := scope.NewClusterScope(context.TODO(), scope.ClusterScopeParams{
+		Client:     fakec,
+		Cluster:    fakeCluster,
+		GCPCluster: fakeGCPCluster,
+		GCPServices: scope.GCPServices{
+			Compute: &compute.Service{},
+		},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name              string
+		scope             func() Scope
+		mockInstanceGroup *cloud.MockInstanceGroups
+		want              []*compute.InstanceGroup
+		wantErr           bool
+	}{
+		{
+			name:  "error getting instanceGroup with non 404 error code (should return an error)",
+			scope: func() Scope { return clusterScope },
+			mockInstanceGroup: &cloud.MockInstanceGroups{
+				ProjectRouter: &cloud.SingleProjectRouter{ID: "proj-id"},
+				Objects:       map[meta.Key]*cloud.MockInstanceGroupsObj{},
+				GetHook: func(ctx context.Context, key *meta.Key, m *cloud.MockInstanceGroups) (bool, *compute.InstanceGroup, error) {
+					return true, &compute.InstanceGroup{}, &googleapi.Error{Code: http.StatusBadRequest}
+				},
+			},
+			want:    []*compute.InstanceGroup{},
+			wantErr: true,
+		},
+		{
+			name:  "instanceGroup does not exist (should create instanceGroup)",
+			scope: func() Scope { return clusterScope },
+			mockInstanceGroup: &cloud.MockInstanceGroups{
+				ProjectRouter: &cloud.SingleProjectRouter{ID: "proj-id"},
+				Objects:       map[meta.Key]*cloud.MockInstanceGroupsObj{},
+			},
+			want: []*compute.InstanceGroup{
+				&compute.InstanceGroup{
+					Name:       "my-cluster-apiserver-us-central1-a",
+					NamedPorts: []*compute.NamedPort{{Name: "apiserver", Port: 6443}},
+					SelfLink:   "https://www.googleapis.com/compute/v1/projects/proj-id/zones/us-central1-a/instanceGroups/my-cluster-apiserver-us-central1-a",
+				},
+			},
+		},
+		{
+			name: "instanceGroup name is overridden (should create instanceGroup)",
+			scope: func() Scope {
+				tagOverride := "master"
+				clusterScope.GCPCluster.Spec.LoadBalancer = infrav1.LoadBalancerSpec{
```

Comment:
Isn't this changing a variable in a shared scope? If the tests are run in a random order, couldn't this set `clusterScope.GCPCluster.Spec.LoadBalancer` for the rest of the suite?

Reply:
Good catch. Changed to use a separate scope to avoid overwriting.

Reply:
Thanks! That'll work.

Another pattern I've seen in tests for other providers is for the function to take in the "golden" one as an argument and return a new copy with the modifications.

Something like  

```go
scope: func(s Scope) Scope {
  s.GCPCluster.Spec.LoadBalancer = infrav1.LoadBalancerSpec{....}
  return s
}
```

and

```go
scope: func(s Scope) Scope {
 return s
}
```

That way, you get to have the top-level one remain clean and passed in on each invocation of `test.scope` to get a fresh, mutable copy.

Up to you if you want to use that here or not.

Reply:
Good suggestion. I've reworked it to make the mods only in the test which saves some duplication.